repose-aggregator/commons/commons-utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletRequestWrapper.scala

```diff
@@ -145,4 +150,81 @@ class HttpServletRequestWrapper(originalRequest: HttpServletRequest, inputStream
   def getSplittableHeaderScala(headerName: String): List[String] = getHeadersScala(headerName).foldLeft(List.empty[String])((list, s) => list ++ s.split(","))
 
   override def getSplittableHeaders(headerName: String): util.List[String] = getSplittableHeaderScala(headerName).asJava
+
+  override def getRequestURL: StringBuffer = requestUrl
+
+  def setRequestURL(url: StringBuffer): Unit = {
+    if (Option(url).isEmpty) throw new IllegalArgumentException("null is not a legal argument to setRequestURL")
+
+    requestUrl = url
+  }
+
+  override def getRequestURI: String = requestUri
+
+  def setRequestURI(uri: String): Unit = {
+    if (Option(uri).isEmpty) throw new IllegalArgumentException("null is not a legal argument to setRequestURI")
+
+    requestUri = uri
+  }
+
+  /**
+    * @return a string representation of the query parameters for this request
+    */
+  override def getQueryString: String = queryString
+
+  /**
+    * @param string the desired query string for this request
+    */
+  def setQueryString(string: String): Unit = {
+    // todo: update parameter map
+    // parse the current query string and remove from parameter map
+    // parse the new query string and add to parameter map
+    queryString = string
+  }
+
+  /**
+    * @param key a parameter key
+    * @return the first parameter value associated with the provided key for this request, or null if no value exists
+    */
+  override def getParameter(key: String): String =
+    Option(getParameterValues(key)).map(_.head).orNull
+
+  /**
+    * @param key a parameter key
+    * @return all parameter values associated with the provided key for this request
+    */
+  override def getParameterValues(key: String): Array[String] =
+    parameterMap.map(_.get(key).orNull).getOrElse(super.getParameterValues(key))
+
+  /**
+    * @return all parameter names for this request
+    */
+  override def getParameterNames: util.Enumeration[String] =
+    parameterMap.map(_.keysIterator.asJavaEnumeration).getOrElse(super.getParameterNames)
+
+  /**
+    * @return the parameter map for this request
+    */
+  override def getParameterMap: util.Map[String, Array[String]] =
+    parameterMap.map(_.asJava).getOrElse(super.getParameterMap)
```

Comment:
This is probably broken. If setParameterMap(...) has not been called, the underlying mutable parameter map is return. It can be mutated, and those mutations will persist. This can be done until setParameterMap(...) is called and the map is made immutable.

One option would be to copy over the mutable map into the immutable map upon construction. That could be costly, and honestly, mutability is kind of baked into the servlet spec.

So instead, we could just allow the parameter map to be mutable. Originally, I thought that the parameter map and query string had the same contents, but that turns out not to be the case. If it were the case, a mutable parameter map would enable de-synchronization between the parameter map and the query string. If we wanted to allow the parameter map to be mutable, I don't believe we would have to override any of the parameter methods (since, again, the underlying map is mutable).

If we take that mutable parameter map approach, what about the query string? Modifications to the parameter map should not be reflected in the query string, since the parameter map is comprised of more than just query parameters. To change the query string, a user should have to use the setQueryString(...) method. However, since the parameter map does include the query string parameters, the setQueryString(...) method should update the parameter map.